rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp

```diff
@@ -0,0 +1,308 @@
+    this->noise_model->max_depth = _sdf->GetElement("depthNoiseMaxDist")->Get<float>();
+  }
+
+  load_connection_ = GazeboRosCameraUtils::OnLoad(boost::bind(&GazeboNoisyDepth::Advertise, this));
+  GazeboRosCameraUtils::Load(_parent, _sdf);
+
+}
+
+void GazeboNoisyDepth::Advertise() {
+
+  ros::AdvertiseOptions depth_image_ao =
+      ros::AdvertiseOptions::create<sensor_msgs::Image>(
+          this->depth_image_topic_name_, 1,
+          boost::bind(&GazeboNoisyDepth::DepthImageConnect, this),
+          boost::bind(&GazeboNoisyDepth::DepthImageDisconnect, this),
+          ros::VoidPtr(), &this->camera_queue_);
+  this->depth_image_pub_ = this->rosnode_->advertise(depth_image_ao);
+
+  ros::AdvertiseOptions depth_image_camera_info_ao =
+      ros::AdvertiseOptions::create<sensor_msgs::CameraInfo>(
+          this->depth_image_camera_info_topic_name_, 1,
+          boost::bind(&GazeboNoisyDepth::DepthInfoConnect, this),
+          boost::bind(&GazeboNoisyDepth::DepthInfoDisconnect, this),
+          ros::VoidPtr(), &this->camera_queue_);
+  this->depth_image_camera_info_pub_ = this->rosnode_->advertise(depth_image_camera_info_ao);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Increment count
+void GazeboNoisyDepth::DepthImageConnect() {
+  this->depth_image_connect_count_++;
+  this->parentSensor->SetActive(true);
+}
+////////////////////////////////////////////////////////////////////////////////
+// Decrement count
+void GazeboNoisyDepth::DepthImageDisconnect() {
+  this->depth_image_connect_count_--;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Increment count
+void GazeboNoisyDepth::DepthInfoConnect() {
+  this->depth_info_connect_count_++;
+}
+////////////////////////////////////////////////////////////////////////////////
+// Decrement count
+void GazeboNoisyDepth::DepthInfoDisconnect() {
+  this->depth_info_connect_count_--;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Update the controller
+void GazeboNoisyDepth::OnNewDepthFrame(const float *_image,
+                                       unsigned int _width, unsigned int _height, unsigned int _depth,
+                                       const std::string &_format) {
+  if (!this->initialized_ || this->height_ <= 0 || this->width_ <= 0)
+    return;
+
+  this->depth_sensor_update_time_ = this->parentSensor->LastMeasurementTime();
+  if (this->parentSensor->IsActive()) {
+    if (this->depth_image_connect_count_ <= 0 &&
+        (*this->image_connect_count_) <= 0) {
+      this->parentSensor->SetActive(false);
+    } else {
+      if (this->depth_image_connect_count_ > 0)
+        this->FillDepthImage(_image);
+    }
+  } else {
+    if (this->depth_image_connect_count_ <= 0)
+      // do this first so there's chance for sensor to run 1 frame after activate
+      this->parentSensor->SetActive(true);
+  }
```

Comment:
This seems a bit complicated, I guess the important part is filling the depth image when `depth_image_connect_count_ > 0 && this->parentSensor->IsActive()` and then you want a functionality do activate or deactivate the parent sensor. (But this part I didn't fully get what you wanted to do)

Reply:
as the logic is somewhat duplicate between OnNewImageFrame and OnNewDepthFrame it's confusing. In one of them it was actually implemented correctly (IMHO). 
The second check should check if there are image subscribers on the parent (inherited property image_connect_count) and if so, activate the parent sensor, before outputing sensor data itself.

Fixed it and commented it a bit. Seems to be ok.